src/Microsoft.Extensions.FileProviders.Physical/PollingWildCardChangeToken.cs

```diff
@@ -0,0 +1,153 @@
+using Microsoft.Extensions.FileSystemGlobbing;
+using Microsoft.Extensions.FileSystemGlobbing.Abstractions;
+using Microsoft.Extensions.Primitives;
+
+namespace Microsoft.Extensions.FileProviders.Physical
+{
+    public class PollingWildCardChangeToken : IChangeToken
+    {
+        public static readonly TimeSpan PollingInterval = TimeSpan.FromSeconds(4);
+        private const int DefaultBufferSize = 256;
+        private readonly object _enumerationLock = new object();
+        private readonly DirectoryInfoBase _directoryInfo;
+        private readonly Matcher _matcher;
+        private bool _changed;
+        private DateTime _lastScanTimeUtc;
+        private byte[] _fileNameBuffer;
+        private byte[] _previousHash;
+
+        public PollingWildCardChangeToken(
+            string root,
+            string pattern)
+            : this(
+                new DirectoryInfoWrapper(new DirectoryInfo(root)),
+                pattern,
+                Internal.Clock.Instance)
+        {
+        }
+
+        public PollingWildCardChangeToken(
+            DirectoryInfoBase directoryInfo,
+            string pattern,
+            IClock clock)
+        {
+            _directoryInfo = directoryInfo;
+            Clock = clock;
+
+            _matcher = new Matcher(StringComparison.OrdinalIgnoreCase);
+            _matcher.AddInclude(pattern);
+            CalculateChanges();
+        }
+
+        /// <inheritdoc />
+        public bool ActiveChangeCallbacks => false;
+
+        public IClock Clock { get; }
+
+        /// <inheritdoc />
+        public bool HasChanged
+        {
+            get
+            {
+                if (_changed)
+                {
+                    return _changed;
+                }
+
+                if (Clock.UtcNow - _lastScanTimeUtc >= PollingInterval)
+                {
+                    lock (_enumerationLock)
+                    {
+                        _changed = CalculateChanges();
+                    }
+                }
+
+                _lastScanTimeUtc = Clock.UtcNow;
+                return _changed;
+            }
+        }
+
+        private bool CalculateChanges()
+        {
+            var result = _matcher.Execute(_directoryInfo);
+            var hasChanges = false;
+
+            var files = result.Files.OrderBy(f => f.Path, StringComparer.OrdinalIgnoreCase);
+#if NET451
+            using (var sha256 = new IncrementalHash())
+#else
+            using (var sha256 = IncrementalHash.CreateHash(HashAlgorithmName.SHA256))
+#endif
+            {
+                foreach (var file in files)
+                {
+                    if (!hasChanges &&
+                        _lastScanTimeUtc < GetLastWriteUtc(file.Path))
+                    {
+                        // _lastScanTimeUtc is the greatest timestamp that any last writes could have been.
+                        // If a file has a newer timestamp than this value, it must've changed.
+                        hasChanges = true;
```

Comment:
It looks fine to me, the hashes for every file are computed and the GetLastWriteUtc check is skipped if any files have changed. Unless you meant we could short circuit here if hasChanges is never true...